cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs

```diff
@@ -93,11 +104,127 @@ import Distribution.Simple.Command
+            rest <- go [] xs
+            pure . fmap mconcat . sequence $ [fs, res, rest]
+         (ParseUtils.Section l "if" p xs') -> do
+                           subpcs <- go [] xs'
+                           let fs = fmap singletonProjectConfigSkeleton $ fieldsToConfig (reverse acc)
+                           (elseClauses, rest) <- parseElseClauses xs
+                           let condNode =  (\c pcs e -> CondNode mempty mempty [CondBranch c pcs e]) <$>
+                                  -- we rewrap as as a section so the readFields lexer of the conditional parser doesn't get confused
+                                  adaptParseError l (parseConditionConfVarFromClause . BS.pack $ "if(" <> p <> ")") <*>
+                                  subpcs <*>
+                                  elseClauses
+                           pure . fmap mconcat . sequence $ [fs, condNode, rest]
+         _ -> go (x:acc) xs
+    go acc [] = pure . fmap singletonProjectConfigSkeleton . fieldsToConfig $ reverse acc
+
+    parseElseClauses :: [ParseUtils.Field] -> IO (ParseResult (Maybe ProjectConfigSkeleton), ParseResult ProjectConfigSkeleton)
+    parseElseClauses x = case x of
+         (ParseUtils.Section _l "else" _p xs':xs) -> do
+               subpcs <- go [] xs'
+               rest <- go [] xs
+               pure (Just <$> subpcs, rest)
+         (ParseUtils.Section l "elif" p xs':xs) -> do
+               subpcs <- go [] xs'
+               (elseClauses, rest) <- parseElseClauses xs
+               let condNode = (\c pcs e -> CondNode mempty mempty [CondBranch c pcs e]) <$>
+                                  adaptParseError l (parseConditionConfVarFromClause . BS.pack $ "else("<> p <> ")") <*>
+                                  subpcs <*>
+                                  elseClauses
+               pure (Just <$> condNode, rest)
+         _ -> (\r -> (pure Nothing,r)) <$> go [] x
+
+    fieldsToConfig xs = fmap (addProvenance . convertLegacyProjectConfig) $ parseLegacyProjectConfigFields source xs
+    addProvenance x = x {projectConfigProvenance = Set.singleton (Explicit source)}
+
+    adaptParseError _ (Right x) = pure x
+    adaptParseError l (Left e) = parseFail $ ParseUtils.FromString (show e) (Just l)
+
+    liftPR :: (a -> IO (ParseResult b)) -> ParseResult a -> IO (ParseResult b)
+    liftPR f (ParseOk ws x) = addWarnings <$> f x
+       where addWarnings (ParseOk ws' x') = ParseOk (ws' ++ ws) x'
+             addWarnings x' = x'
+    liftPR _ (ParseFailed e) = pure $ ParseFailed e
+
+    fetchImportConfig :: ProjectConfigImport -> IO BS.ByteString
+    fetchImportConfig pci = case parseURI pci of
+         Just uri -> do
+            let fp = cacheDir </> map (\x -> if isPathSeparator x then 'X' else x) (show uri) -- TODO can we do better?
+            createDirectoryIfMissing True cacheDir
+            _ <- downloadURI httpTransport verbosity uri fp
```

Comment:
I'm not sure from this piece of code if and how these remote files are cached. I think we discussed that wrt remote freeze files somewhere and "cache strategy" here: https://github.com/haskell/cabal/issues/6528#issuecomment-851846895

Reply:
This uses the same caching strategy used by cabal's caching of metadata from hackage, etc. which attempts to use built in http features with the `if-modified-since` header and etags.

Reply:
I think that might still be excessive in some cases, e.g. when using `cabal list-bins`.

In ghcup, relying only on etags etc causes a considerable slow down as well of 200-300ms (much more when using wget). There I use an additional  5minute period after last file access to determine whether to invoke any http methods.

Not sure that's the right thing here. Maybe we do want this configurable.